go/test/endtoend/vtgate/queries/kill/kill_test.go

```diff
@@ -0,0 +1,234 @@
+	// TODO: does not really change anything, but expect to receive Queery Interrupted error
+	//  "(errno 1317) (sqlstate 70100)"
+	require.NoError(t, err)
+}
+
+// TestKillDifferentConnectionQuery kills query on different connection and check relevant error messages.
+func TestKillDifferentConnectionQuery(t *testing.T) {
+	setupData(t, false)
+	defer dropData(t)
+
+	conn, err := mysql.Connect(context.Background(), &vtParams)
+	require.NoError(t, err)
+	defer conn.Close()
+
+	killConn, err := mysql.Connect(context.Background(), &vtParams)
+	require.NoError(t, err)
+	defer killConn.Close()
+
+	// killing query on non-existent connection
+	_, err = utils.ExecAllowError(t, killConn, "kill query 12345")
+	require.ErrorContains(t, err, "Unknown thread id: 12345 (errno 1094) (sqlstate HY000)")
+
+	done := make(chan error)
+	go func() {
+		// 20 seconds sleep. Should be stopped by kill statement.
+		_, err := utils.ExecAllowError(t, conn, "select sleep(20) from test")
+		done <- err
+	}()
+
+	for {
+		select {
+		case execErr := <-done:
+			require.ErrorContains(t, execErr, "context canceled (errno 1317) (sqlstate 70100)")
+			return
+		case <-time.After(100 * time.Millisecond):
+			_, err = utils.ExecAllowError(t, killConn, fmt.Sprintf("kill query %d", conn.ConnectionID))
+			require.NoError(t, err)
+		case <-time.After(5 * time.Second):
+			t.Fatal("test did not complete in 5 seconds.")
+		}
+	}
+}
+
+// TestKillOnHungQuery test that any hung query should return.
+func TestKillOnHungQuery(t *testing.T) {
+
+	t.Run("connection close", func(t *testing.T) {
+		testHungQuery(t, func(conn1 *mysql.Conn, conn2 *mysql.Conn) {
+			// closing the hung query connection.
+			conn2.Close()
+		}, "(errno 2013) (sqlstate HY000)")
+	})
+
+	t.Run("connection kill", func(t *testing.T) {
+		testHungQuery(t, func(conn1 *mysql.Conn, conn2 *mysql.Conn) {
+			// kill the hung connection
+			utils.Exec(t, conn1, fmt.Sprintf("kill %d", conn2.ConnectionID))
+		}, "context canceled")
+	})
+
+	t.Run("query kill", func(t *testing.T) {
+		testHungQuery(t, func(conn1 *mysql.Conn, conn2 *mysql.Conn) {
+			// kill the hung query
+			utils.Exec(t, conn1, fmt.Sprintf("kill query %d", conn2.ConnectionID))
+		}, "context canceled")
+	})
+}
+
+func testHungQuery(t *testing.T, killFunc func(*mysql.Conn, *mysql.Conn), errMsg string) {
+	conn1, err := mysql.Connect(context.Background(), &vtParams)
+	require.NoError(t, err)
+	defer conn1.Close()
+
+	utils.Exec(t, conn1, "begin")
+	utils.Exec(t, conn1, "insert into test(id, msg, extra) values (1, 'a', 'e')")
+
+	conn2, err := mysql.Connect(context.Background(), &vtParams)
+	require.NoError(t, err)
+	defer conn2.Close()
+
+	ch := make(chan error)
+	go func() {
+		ch <- nil
+		utils.Exec(t, conn2, "begin")
+		_, err := utils.ExecAllowError(t, conn2, "insert into test(id, msg, extra) values (1, 'a', 'e')")
+		ch <- err
+	}()
+	<-ch                        // started
+	time.Sleep(1 * time.Second) // just little more time
```

Comment:
this channel thingie here is difficult to follow. you are using it both to control the flow of the goroutine but also for the error output

Reply:
`testHungQuery` and `testHugeData`, similar logic with different way of waiting. unifying them now.